src/lib/hqueue.rs

```diff
@@ -0,0 +1,296 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+
+use std::cmp::Ordering;
+use std::fmt;
+
+use ast::{Arena, NodeId};
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+/// A `PriorityNodeId` wraps the height of a node with its id.
+///
+/// This type should be completely opaque to clients of this module.
+/// Client code should construct a `HeightQueue` and call its methods,
+/// which will return `NodeId`s directly, rather than the `PriorityNodeId`
+/// wrapper.
+struct PriorityNodeId {
+    index: NodeId,
+    height: u32,
+}
+
+impl PriorityNodeId {
+    fn new(index: NodeId, height: u32) -> PriorityNodeId {
+        PriorityNodeId {
+            index: index,
+            height: height,
+        }
+    }
+
+    fn id(&self) -> NodeId {
+        self.index
+    }
+
+    fn height(&self) -> u32 {
+        self.height
+    }
+}
+
+impl Ord for PriorityNodeId {
+    fn cmp(&self, other: &PriorityNodeId) -> Ordering {
+        self.height.cmp(&other.height)
+    }
+}
+
+impl PartialOrd for PriorityNodeId {
+    fn partial_cmp(&self, other: &PriorityNodeId) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+/// A queue of `NodeId`s sorted on the height of their respective nodes.
+pub struct HeightQueue {
+    queue: Vec<PriorityNodeId>, // Use Vec so we can call `sort()`.
+}
+
+impl fmt::Debug for HeightQueue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[ ")?;
+        for item in &self.queue {
+            write!(f, "({:?}, {:?}) ", item.id(), item.height())?;
+        }
+        write!(f, "]")
+    }
+}
+
+impl HeightQueue {
+    /// Create empty priority queue.
+    pub fn new() -> HeightQueue {
+        HeightQueue { queue: Vec::new() }
+    }
+
+    /// Remove (and discard) all items in this queue, leaving it empty.
+    pub fn clear(&mut self) {
+        self.queue.clear();
+    }
+
+    /// `true` if this queue is empty, `false` otherwise.
+    pub fn is_empty(&self) -> bool {
+        self.queue.is_empty()
+    }
+
+    /// Get the id of the `Node` with the greatest height in the current queue.
+    pub fn peek_max(&self) -> Option<NodeId> {
+        if self.queue.is_empty() {
+            return None;
+        }
+        return Some(self.queue[self.queue.len() - 1].index);
+    }
+
+    /// Remove information about the tallest node and return its `NodeId`.
+    pub fn pop(&mut self) -> Option<NodeId> {
+        if self.is_empty() {
+            return None;
+        }
+        Some(self.queue.pop().unwrap().index)
+    }
+
+    /// Push a new node into this priority queue, keeping the queue sorted.
+    ///
+    /// This method has no effect if the new node is already in the queue.
+    /// This method is potentially slow, use `push_many()` to add a number
+    /// of new items to the queue at once.
```

Comment:
How many items deep is the queue likely to get? If it's more than ~16 this method is going to become *really* slow as it's O(n^2) in the worst case. You could make it a linked list at which point it'd become O(n) (but obviously there are other costs with linked lists); or you could replace case 4 with a binary chop which would make it O(n log n). But it's only worth thinking about this if the queue is going to become quite deep.